dotnet/src/InternalUtilities/process/StepExtensions.cs

```diff
@@ -0,0 +1,166 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.SemanticKernel.Process.Runtime;
+
+internal static class StepExtensions
+{
+    /// <summary>
+    /// The generic state type for a process step.
+    /// </summary>
+    private static readonly Type s_genericType = typeof(KernelProcessStep<>);
+
+    public static KernelProcessStepInfo Clone(this KernelProcessStepInfo step, ILogger logger)
+    {
+        if (step is KernelProcess subProcess)
+        {
+            return subProcess.CloneProcess(logger);
+        }
+
+        Type stateType = step.InnerStepType.InspectStateType(out Type? userStateType, logger);
+
+        KernelProcessStepState newState = step.State.Clone(stateType, userStateType);
+
+        KernelProcessStepInfo copy =
+            new(
+                step.InnerStepType,
+                newState,
+                step.Edges.ToDictionary(kvp => kvp.Key, kvp => kvp.Value.ToList()));
+
+        return copy;
+    }
+
+    // Exposed for testing
+    public static KernelProcessStepState Clone(this KernelProcessStepState sourceState, Type stateType, Type? userStateType)
+    {
+        KernelProcessStepState newState = (KernelProcessStepState)Activator.CreateInstance(stateType, sourceState.Name, sourceState.Id)!; // %%% EXCEPTION / NULL RESULT
```

Comment:
Is the comment `// %%% EXCEPTION / NULL RESULT` still meaningful?

Reply:
Right, good call-out...I just want to do an explicit null check instead of `!`...updating now